cluster-autoscaler/simulator/drainability/rules/replicacount/rule.go

```diff
@@ -57,7 +58,10 @@ func (r *Rule) Drainable(drainCtx *drainability.DrainContext, pod *apiv1.Pod) dr
 	if controllerRef == nil {
 		return drainability.NewUndefinedStatus()
 	}
-	refKind := controllerRef.Kind
+
+	groupVersionKind := schema.FromAPIVersionAndKind(controllerRef.APIVersion, controllerRef.Kind)
```

Comment:
Docs for `FromAPIVersionAndKind` says:
```go
// FromAPIVersionAndKind returns a GVK representing the provided fields for types that
// do not use TypeMeta. This method exists to support test types and legacy serializations
// that have a distinct group and kind.
// TODO: further reduce usage of this method.
```
https://github.com/kubernetes/apimachinery/blob/e2f405af78de67e27b5dd5ceb1a3eca76331e4d5/pkg/runtime/schema/group_version.go#L296-L305

`DaemonSet` uses `TypeMeta`: https://pkg.go.dev/k8s.io/api/apps/v1#DaemonSet

Should we be using `TypeMeta`'s [`GroupVersionKind()`](https://github.com/kubernetes/apimachinery/blob/v0.29.0/pkg/apis/meta/v1/meta.go#L125) function here instead? `GroupVersionKind()` uses [`FromAPIVersionAndKInd()` internally](https://github.com/kubernetes/apimachinery/blob/3c8c1f22dc332a8a6b51033f6a7c0566b885d290/pkg/apis/meta/v1/meta.go#L125-L127) but there is less chance of `GroupVersionKind()` getting removed in the future (while `FromAPIVersionKind()` might)  

Something like (at the cost of more verbosity):
```go
v1.TypeMeta{Kind: controllerRef.Kind, APIVersion: controllerRef.APIVersion}.GroupVersionKind()
```

Reply:
cc: @x13n

Reply:
How about:

```
refGroup := schema.ParseGroupVersion(controllerRef.APIVersion)
refKind := controllerRef.Kind
```

`ParseGroupVersion` isn't annotated as something that is planned for removal.

Reply:
Sounds good to me :+1:

Reply:
Hi @vladimirvivien , @x13n 

`schema.ParseGroupVersion` returns error in addition to `schema.GroupVersion`, will the following code be sufficient?

```go
refGroup, err := schema.ParseGroupVersion(controllerRef.APIVersion)
if err != nil {
	return drainability.NewUndefinedStatus()
}

refKind := controllerRef.Kind
```

Or I should return `drainability.NewBlockedStatus` instead?

Reply:
Undefined is ok in this case. If it didn't parse, it means it wasn't a correctly formatted `APIVersion` string.

Reply:
Allright, changed